[Detection Engine][Exceptions] - Update docs for single and shared exceptions (#4021)
elastic/security-docs: docs/detections/api/rules/rules-api-create-rule-default-exception-list.asciidoc

[[exceptions-api-create-rule-default-exception-list]]

> === Create a default exception list for a rule

Author: ```suggestion
=== Create default exception list for a rule
```

> Creates a default exception list for the rule you specify. Default exception lists can only be associated with a single rule.

Reviewer: My wording isn't a great, but maybe something like:

Adds specified exception items to a rules default exception list - if no default exception list exists, one is created and associated with the rule. Default exception lists are meant to house exceptions associated with a single rule, as opposed to shared exception lists which can be associated to multiple rules.

> Creates a default exception list for the rule you specify. Default exception lists can only be associated with a single rule.

Reviewer 2: I think the second sentence of this first paragraph would make more sense in the middle of the second paragraph, and the first sentence of the first paragraph would make more sense without it adjacent, too.

As an alternative, you also making the NOTE in this section a bit bigger by including this sentence as an additional point

Author: Yeah, I can see how these sentences repeat the same message. I like your suggestion to add the second sentence to the note. It'd be a good reminder why default exception lists can't be accessed from the Shared Exception Lists page.

To add exception items to a default exception list, pass in exceptions items that you want applied to the rule. Refer to <<exceptions-api-create-rule-default-exception-item>> for more information. 

> When an exception item’s query evaluates to `true`, the associated rule does not issue alerts even when its other criteria are met.

Reviewer 2: This might also fit within the NOTE if you decide to expand it as my previous comment suggested. Otherwise it might be okay to take it out — hopefully someone looking at this already knows what rule exceptions are for. 

I think if you're going to include some background info here, it might be nice to assuage any confusion on the part of the reader about the interchangeable use of "exception list" and "exception container".

Author: I still think line 8 has value for users who are new to exceptions, so I don't want to remove it. 

RE your observation about the interchangeable use of "exception list" and "exception container": I replaced every instance of "exception container" with "exception list" on this page so there's consistency in the terminology.

> NOTE: Default exception lists do not display on the <<shared-exception-lists, Shared Exception Lists>> page in the {security-app} UI. 

Reviewer: Should we add where they do show up? Just in the rule details, exceptions tab?

Author: Added in [585ed2c](https://github.com/elastic/security-docs/pull/4021/commits/585ed2cb75a83d00203852718b0a7d5fb610f4f6).

> NOTE: Default exception lists do not display on the <<shared-exception-lists,Shared Exception Lists>> page in the {security-app} UI. 

Author: ```suggestion
NOTE: Default exception lists do not display on the <<shared-exception-lists, Shared Exception Lists>> page in the {security-app} UI. 
```

> NOTE: Default exception lists do not display on the **Shared Exception Lists page in the {security-app} UI. They only appear in the Rule exceptions on the rule details page. Refer to <<manage-exception>> to learn more.

Reviewer 2: ```suggestion
NOTE: Default exception lists do not display on the **Shared Exception Lists** page in the {security-app} UI. They only appear in the Rule exceptions on the rule details page. Refer to <<manage-exception>> to learn more.
```

Author: Nice catch - thanks!

==== Request URL

`POST <kibana host>:<port>/api/detection_engine/<rule_id>/exceptions`

==== Request body

A JSON object with these fields:

[width="100%",options="header"]
|==============================================
|Name |Type |Description |Required

|`description` |String |Describes the exception container. |Yes
|`list_id` |String |Unique identifier. |No, automatically created when it is not
provided.
|`meta` |Object |Placeholder for metadata about the list container. |No
|`name` |String |The exception container's name. |Yes
|`namespace_type` |String a|Determines whether the exception container is available in all {kib} spaces or just the space in which it is created, where:

* `single`: Only available in the {kib} space in which it is created.
* `agnostic`: Available in all {kib} spaces.

|No, defaults to `single`.
|`tags` |String[] |String array containing words and phrases to help categorize
exception containers. |No
|`type` |String a|The type of exception, which must be:

> * `rule_default`: Exception list that belongs to a single rule

Reviewer 2: ```suggestion
* `rule_default`: Exception list that belongs to a single rule.
```

|Yes

|==============================================

===== Example requests

Creates an exception container for holding trusted Linux process exception
items:

[source,console]
--------------------------------------------------

> POST api/exception_lists

Reviewer: Thanks so much for adding these docs! I'm wondering if this page is needed - we really just want users to utilize the `POST /api/detection_engine/rules/<id>/exceptions` endpoint and specify items they want added to a rule. The idea with that endpoint was to not have users need to worry about creating the "parent" or "container" part of the exception list. 

I also don't think this example matches the endpoint - /api/detection_engine/<rule_id>/exceptions.

Author: Hm, good question. I'm unsure whether public API endpoints _always_ need to be documented, especially when another endpoint does the same thing but in a slightly different way. Is there a way to determine whether any of our users are using this endpoint? Also, do you know if this endpoint is still being maintained and if it will be updated in the future?

Author: Here's the fixed path in case we keep this page:

```suggestion
POST api/detection_engine/<rule_id>/exceptions
```

Author: @yctercero I checked with one of our API writers, and the general advice is that public endpoints should be doc'd so we should prob keep this page.

Author: This might no longer be an issue since we're directing users towards the new reference site, which will/should have all the latest API docs for public Security endpoints.

{
  "description": "Excludes Linux trusted processes",
  "name": "Linux process exceptions",
  "list_id": "trusted-linux-processes",
  "type": "detection",
  "namespace_type": "single",
  "tags": [
    "linux",
    "processes"
  ]
}
--------------------------------------------------
// KIBANA

==== Response code

`200`::
    Indicates a successful call.

==== Response payload

The exception container object with a unique ID.

[source,json]
--------------------------------------------------
{
  "_tags": [],
  "created_at": "2020-07-13T09:33:46.187Z",
  "created_by": "elastic",
  "description": "Excludes Linux trusted processes",
  "id": "f320c070-c4eb-11ea-80bb-11861bae2798", <1>
  "list_id": "trusted-linux-processes", <2>
  "name": "Linux process exceptions",
  "namespace_type": "single", <3>
  "tags": [
    "linux",
    "processes"
  ],
  "tie_breaker_id": "2c08d5a5-2ecc-4d5a-acfb-0a367f25b3f3",
  "type": "detection", <4>
  "updated_at": "2020-07-13T09:33:46.359Z",
  "updated_by": "elastic"
}
--------------------------------------------------

> These values are required to associate the exception container with detection
> rules:

Reviewer 2: ```suggestion
The highlighted values can help you identify detection rules connected to the exception container:
```

Author: Fixed in [68d3cb3](https://github.com/elastic/security-docs/pull/4021/commits/68d3cb3078a6526ae5bc8b625905cd7fa2110bfa).
